libcudacxx/include/cuda/std/inplace_vector

```diff
@@ -0,0 +1,1808 @@
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_reference
+  operator[](const size_type __count) const noexcept
+  {
+    return *(this->begin() + __count);
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr reference front() noexcept
+  {
+    return *this->begin();
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_reference front() const noexcept
+  {
+    return *this->begin();
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr reference back() noexcept
+  {
+    return *(this->end() - 1);
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_reference back() const noexcept
+  {
+    return *(this->end() - 1);
+  }
+
+  using __base::data;
+
+  // inplace_vector.iterators
+  using __base::begin;
+  using __base::end;
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr reverse_iterator rbegin() noexcept
+  {
+    return reverse_iterator{this->end()};
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_reverse_iterator rbegin() const noexcept
+  {
+    return const_reverse_iterator{this->end()};
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr reverse_iterator rend() noexcept
+  {
+    return reverse_iterator{this->begin()};
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_reverse_iterator rend() const noexcept
+  {
+    return const_reverse_iterator{this->begin()};
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_iterator cbegin() const noexcept
+  {
+    return this->begin();
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_iterator cend() const noexcept
+  {
+    return this->end();
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_reverse_iterator crbegin() const noexcept
+  {
+    return const_reverse_iterator{this->end()};
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_reverse_iterator crend() const noexcept
+  {
+    return const_reverse_iterator{this->begin()};
+  }
+
+  // [containers.sequences.inplace.vector.members] size/capacity
+  using __base::empty;
+  using __base::size;
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr size_type max_size() const noexcept
+  {
+    return _Capacity;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr size_type capacity() const noexcept
+  {
+    return _Capacity;
+  }
+
+  // [containers.sequences.inplace.vector.modifiers], modifiers
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr iterator insert(const_iterator __cpos, const _Tp& __value)
+  {
+    return emplace(__cpos, __value);
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr iterator insert(const_iterator __cpos, _Tp&& __value)
+  {
+    return emplace(__cpos, _CUDA_VSTD::move(__value));
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr iterator
+  insert(const_iterator __cpos, const size_type __count, const _Tp& __value)
+  {
+    const iterator __pos = (iterator) __cpos;
```

Comment:
That only works with pointers and we might want to move to checked iterators

Reply:
The "most guaranteed" way seems to be
```suggestion
    const iterator __pos = begin() + (__cpos - cbegin());
```

The standard doesn't guarantee to provide a way to obtain an `iterator` from a `const_iterator` without knowing the container (see [[container.reqmts]](https://eel.is/c++draft/container.reqmts)).
But since we're the implementation here, perhaps the `iterator` can be obtained in a more clever way.

Reply:
Yeah, I am a tiny bit torn here.

The obviously correct way is to create a static member function that does this and centralize the evilness.

On the other hand this will just worsen compiler throughput and debug experience

Reply:
Alright, let's leave the cast as is then. Thx for the explanation!

Reply:
Oh i just replaced it with the "correct" solution in https://github.com/NVIDIA/cccl/pull/1936/commits/12d640f55aa3f664fafb2b6d995b0de7ff44dc98

Reply:
Alright, works for me!